### PR TITLE
JAMES-2201 Switch to SHA-256 for attachments indexing and Cassandra blob IDs

### PR DESCRIPTION
**Pull request by thithib**
This is in reaction to rising threats of SHA-1 collisions (see SHAttered
from Google: https://shattered.io/).

### CODE REVIEW

**Comment by chibenwa** [comment]
Could you please give  a lin for the reference you give?

**Comment by thithib** [comment]
From what I understand, Cassandra BlobIds are a different problem and I am still investigating on that.

**Comment by chibenwa** [comment]
You can very well imagine creating conflict based on that.

To reproduce:
 - Send a first mail with only (and only) an attachment being the PDF from SHAttered (No Mime multipart) 
 - Then send a second with the other PDF (with same structure). You should notice the similar fail.

You can also write test at the `unit test` level to reproduce that. IE it is also vulnerable to hash collision.

**Comment by thithib** [comment]
Yes, I intended to fix that in another PR when I have fully understood the issue, but I will rather add another commit to this PR later if you prefer.

**Comment by chibenwa** [comment]
test this please

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
I would rely love see the collision case before the fix

**Comment by thithib** [comment]
@mbaechler Download the two PDFs from [SHAttered.io](https://shattered.io/) and send them to yourself in two different emails, for instance with the blue one in the first email and the red one in the second email. Then open the two emails you just received from yourself, download the attachments and you should get the red PDF in both emails.

Test this on OpenPaaS, at it only happens with JMAP because of the way attachments are indexed and stored separately.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/ids/BlobId.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
 public class BlobId {
     public static BlobId forPayload(byte[] payload) {
         Preconditions.checkArgument(payload != null);
-        return new BlobId(DigestUtils.sha1Hex(payload));
+        return new BlobId(DigestUtils.sha256Hex(payload));
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/ids/BlobId.java]: On this PR you have to make three changes for changing the hash used.
I think in a perfect word we should have a static method called "hashForFile" for example that hide the choosen implementation.

**Comment by mbaechler** [comment]
@thithib : I understand very well the test case, what I mean is I would like to see an automatic test added to the test suite

**Pull request merged**